### PR TITLE
Device activation flow with SMS-based agent onboarding

### LINKED ISSUE #86: BE-02 : Activation Code Generation & SMS Integration
# 🎫 BE-02 — Activation Code Generation & SMS Integration

**Priority:** P1  
**Estimate:** 2 days  

---

## Description

Implement SMS-based activation flow for admin-created agents.

When a new agent is provisioned, the backend must generate a secure activation code and send it via SMS.

---

## Acceptance Criteria

- 6-digit numeric activation code.
- Code stored hashed.
- Expiration: 10 minutes.
- Max 5 attempts.
- Marked as used after success.
- SMS sending abstracted behind trait.
- SMS mocked in development.

---

## Implementation Notes

- Use secure random generator.
- Hash activation code before storage.
- Enforce retry limits.

---

## Activity

- Implement activation service.
- Integrate SMS provider.
- Add expiration validation.
- Add unit tests.

### PR DESCRIPTION
**Pull request by Ayonta-2**
**Changes Overview**
Device activation service - Secure token generation and validation for agent device registration
SMS provider integration - Abstraction layer for sending activation codes via SMS
Redis integration - Token storage, expiration handling, and rate limiting
Activation test suite - Comprehensive tests for token generation, validation, and edge cases
CI/CD improvements - Self-hosted runners and code coverage reporting (60% threshold)

**Key Features**
Secure device pairing - 6-digit activation tokens with configurable expiration
Rate limiting - Prevents abuse with attempt counters and cooldown periods
Agent status management - PENDING_ACTIVATION → ACTIVE workflow
SMS abstraction - Pluggable provider interface for different SMS gateways

### CODE REVIEW

File: iviss-backend/src/services/activation_service.rs [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+            serde_json::from_str(&raw).context("Failed to deserialize activation entry")?;
+
+        // Check number of attempts
+        if entry.attempts >= MAX_ATTEMPTS {
+            // Delete the key
+            conn.del::<_, ()>(&key).await.ok();
+            return Err(anyhow::anyhow!(
+                "Max attempts reached — activation code invalidated"
+            ));
+        }
```
Comment (Arielpetit) [inline iviss-backend/src/services/activation_service.rs:104]: There’s an off-by-one issue in the attempts logic.

Currently, you check `entry.attempts >= MAX_ATTEMPTS` first, then increment before verifying. This causes:

* The 5th wrong attempt to set `attempts = 5` but not delete the key.
* The key only gets deleted on the 6th call.

**Expected:**
With “Max 5 attempts”, the key should be invalidated immediately on the 5th failed attempt.

**Suggested fix:**

* Increment `attempts` only after confirming a mismatch.
* If `attempts` reaches 5 after increment → delete the key immediately and return “locked”.

File: iviss-backend/src/services/activation_service.rs [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+        let value =
+            serde_json::to_string(&entry).context("Failed to serialize activation entry")?;
+
+        let mut conn = self
+            .redis
+            .get()
+            .await
+            .context("Failed to get Redis connection")?;
+
+        // Store with TTL 10 minutes — replaces any existing entry
```
Comment (Arielpetit) [inline iviss-backend/src/services/activation_service.rs:61]: TTL is being reset on every attempt.

In `activation_service.rs`, `set_ex(..., ACTIVATION_TTL_SECS)` reapplies 600s each time, making it “10 min since last attempt” instead of “10 min since issuance”.

**Expected:**
Expire 10 minutes after generation, regardless of retries.

**Suggested fix:**

* Read the current TTL.
* Update the value using the remaining TTL (not `ACTIVATION_TTL_SECS`).
* Ideally, switch to an atomic structure to handle this safely.

File: iviss-backend/src/services/activation_service.rs [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+        let code: u32 = rng.gen_range(0..=999_999);
+        format!("{:06}", code) // zero-padded
+    }
+
+    /// SHA-256 hash of the plain code
+    pub(crate) fn hash_code(code: &str) -> String {
+        let mut hasher = Sha256::new();
+        hasher.update(code.as_bytes());
+        format!("{:x}", hasher.finalize())
+    }
```
Comment (Arielpetit) [inline iviss-backend/src/services/activation_service.rs]: Plain SHA-256 on a 6-digit code is brute-forceable.
Since the code space is only 1,000,000 possibilities, a leaked Redis instance allows fast brute-forcing.

**Suggested fix:**
Use a peppered hash :

* **Preferred:** `HMAC-SHA256(pepper, code)`
* **Acceptable:** `SHA256(pepper + ":" + code)`

Store the pepper in config/env (e.g. `ACTIVATION_CODE_PEPPER`) — never in Redis.

**Review by Arielpetit — CHANGES_REQUESTED** [review]

**Comment by Arielpetit** [comment]
Have you added redis in the compose file?

**Review by Arielpetit — APPROVED** [review]
